assets/model_monitoring/components/tests/unit/test_numerical_data_drift_metrics.py

```diff
@@ -1,87 +1,413 @@
+        """
+        Compute the normalized Wasserstein distance between two arrays of numerical data.
 
-        x_pd_df = pd.DataFrame(data=x, columns=column_values)
-        y_pd_df = pd.DataFrame(data=x, columns=column_values)
+        :param x_array: A numpy array of the first distribution's data.
+        :param y_array: A numpy array of the second distribution's data.
+        :return: The normalized Wasserstein distance between the two distributions.
+        """
+        norm = max(np.std(x_array), 0.001)
+        return wasserstein_distance(x_array, y_array) / norm
 
-        spark = init_spark()
+    # # # EXPECTED JENSEN-SHANNON DISTANCE # # #
 
-        x_df = spark.createDataFrame(x_pd_df)
-        y_df = spark.createDataFrame(y_pd_df)
+    def jensen_shannon_distance_numerical(self, x_array, y_array):
+        """
+        Compute the Jensen-Shannon distance between two arrays of numerical data.
 
-        output_df = compute_numerical_data_drift_measures_tests(
-                x_df,
-                y_df,
-                x_df.count(),
-                y_df.count(),
-                "NormalizedWassersteinDistance",
-                column_values,
-                numerical_threshold)
+        :param x_array: A numpy array of the first distribution's data.
+        :param y_array: A numpy array of the second distribution's data.
+        :return: The Jensen-Shannon distance between the two distributions.
+        """
+        bin_edges = self.compute_optimal_histogram_bin_edges(x_array, y_array)
+        x_percent = np.histogram(x_array, bins=bin_edges)[0] / len(x_array)
+        y_percent = np.histogram(y_array, bins=bin_edges)[0] / len(y_array)
 
-        metric_value = self.get_metric_value(output_df, "NormalizedWassersteinDistance")
-        self.assertAlmostEqual(0.0, metric_value, 4)
+        return distance.jensenshannon(x_percent, y_percent, base=2)
 
-    def test_compute_numerical_data_drift_metrics_normalized_wasserstein_distance(self):
-        """Test compute normalized wasserstein distance for numerical metrics."""
-        std_dev = 15
+    # # # EXPECTED PSI # # #
+
+    def psi_numerical(self, x_array, y_array):
+        """
+        Compute the Population Stability Index (PSI) between two arrays of numerical data.
+
+        :param x_array: A numpy array of the first distribution's data.
+        :param y_array: A numpy array of the second distribution's data.
+        :return: The PSI between the two distributions.
+        """
+        bin_edges = self.compute_optimal_histogram_bin_edges(x_array, y_array)
+
+        x_count = np.histogram(x_array, bins=bin_edges)[0]
+        y_count = np.histogram(y_array, bins=bin_edges)[0]
+
+        # Laplace smoothing (incrementing the count of each bin by 1)
+        # to avoid zero values in bins and have the SPI value be finit
+        x_count = x_count + 1
+        y_count = y_count + 1
+
+        x_percent = x_count / len(x_array)
```

Comment:
Do we need to increment the denominator too so that the percentage numbers still add up to 100?

Reply:
I notice the psi metrics are calculated in the same way in the source code, so I guess this is how the Laplace smoothing works. Is that right?

Reply:
This as indeed an issue. It is easy to overlook as the code for categorical data looks like this:
```python
        # Laplace smoothing (incrementing the count of each bin by 1)
        # to avoid zero values in bins and have the SPI value be finit
        x_counts = x_counts + 1
        y_counts = y_counts + 1

        x_ratios = x_counts / np.sum(x_counts)
        y_ratios = y_counts / np.sum(y_counts)
 ```
Which is very similar in structure to the erroneous expression for the numerical data, but logically different. Here we are already normalizing using the updated counts.